File: Qytetet_Ruby/lib/calle.rb
# encoding:UTF-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative "casilla"
require_relative "titulo_propiedad"

module ModeloQytetet
  class Calle < Casilla

    attr_accessor :numHoteles, :numCasas, :titulo
    
    def initialize(numeroCasilla, coste, titulo)
      super(numeroCasilla, coste, TipoCasilla::CALLE)
      @titulo = titulo
      titulo.casilla = self
      @numHoteles = 0
      @numCasas = 0
    end

    def esta_hipotecada
      return @titulo.hipotecada
    end
        
    def get_precio_edificar()
      @titulo.precio_edificar
    end
        
    def edificar_casa()
      @numCasas=@numCasas + 1
      return titulo.precio_edificar
    end
        
    def precio_alquiler
      coste_total = @titulo.alquiler_base
      coste_total += @numCasas * 0.5 + @numHoteles * 2;
      return coste_total;
    end

    def cobrar_alquiler
      coste_total = @titulo.alquiler_base
      coste_total += @numCasas * 0.5 + @numHoteles * 2;
      @titulo.cobrar_alquiler(coste_total);
      return coste_total;
    end

    def tengo_propietario()
      return @titulo.tengo_propietario
    end
          
    def asignar_propietario(jugador)
      @titulo.propietario = jugador;
      return titulo;
    end
          
    def se_puede_edificar_casa (factor)
      return @numCasas < (factor * 4)
    end
          
    def calcular_valor_hipoteca
      hipoteca_base = @titulo.hipoteca_base
      valor =  hipoteca_base + (@numCasas * 0.5 * hipoteca_base + @numHoteles * hipoteca_base);
      return valor;
    end


    def hipotecar
      @titulo.hipotecada = true
      cantidad_recibida = calcular_valor_hipoteca        
      return cantidad_recibida;
    end
          
    def preopietario_encarcelado()
      return @titulo.propietario_encarcelado
    end
          
    def vender_titulo()
      
      @titulo.propietario = nil;
      precio_compra = @numCasas + @numHoteles
      precio_compra = precio_compra * @titulo.precio_edificar
      precio_compra = precio_compra + @coste
      #precio_compra = @coste + (@nunCasas + @numHoteles) * @titulo.precio_edificar;
      precio_venta = precio_compra + @titulo.factor_revalorizacion * precio_compra
      @numCasas = 0
      @numHoteles = 0
      return precio_venta
    end
          
    def se_puede_edificar_hotel(factor)
            
      return @numCasas == (4 * factor)
    end
          
    def edificar_hotel()
      @numCasas = @numCasas - 4
      return titulo.precio_edificar
    end
          
    def get_coste_hipoteca
      return calcular_valor_hipoteca * 1.1
    end
          
    def cancelar_hipoteca()
      valor_cancelar = get_coste_hipoteca
      titulo.hipotecada = false
            
      return valor_cancelar
    end
        
        
=begin   
    
def precio_total_comprar()
    
end    
=end
    def asignar_titulo_propiedad()
            
    end
    private :titulo=,:asignar_titulo_propiedad

  end
end

File: QytetetJava/src/modeloqytetet/Calle.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

/**
 *
 * @author jo_se
 */
public class Calle extends Casilla {

    private int numHoteles;
    private int numCasas;
    private TituloPropiedad titulo;

    public Calle(int numeroCasilla, int coste, TituloPropiedad titulo) {
        super(numeroCasilla, coste, TipoCasilla.CALLE);
        this.titulo = titulo;
        this.numHoteles = 0;
        this.numCasas = 0;
        titulo.setCasilla(this);
    }

    void setNumHoteles(int numHoteles) {
        this.numHoteles = numHoteles;
    }

    void setNumCasas(int numCasas) {
        this.numCasas = numCasas;
    }

    private void setTitulo(TituloPropiedad titulo) {
        this.titulo = titulo;
    }

    int getNumHoteles() {
        return numHoteles;
    }

    int getNumCasas() {
        return numCasas;
    }

    @Override
    public TituloPropiedad getTitulo() {
        return titulo;
    }

    TituloPropiedad asignarPropietario(Jugador jugador) {
        titulo.setPropietario(jugador);
        return titulo;
    }

    int cancelarHipoteca() {
        int valorCancelar = getCosteHipoteca();
        titulo.setHipotecada(false);
        return valorCancelar;
    }

    int calcularValorHipoteca() {
        int hipotecaBase = titulo.getHipotecaBase();
        int valor = hipotecaBase + (int) (numCasas * 0.5 * hipotecaBase + numHoteles * hipotecaBase);

        return valor;
    }

    int cobrarAlquiler() {
        int costeTotal = titulo.getAlquilerBase();
        costeTotal += numCasas * 0.5 + numHoteles * 2;
        titulo.cobrarAlquiler(costeTotal);

        return costeTotal;
    }

    int edificarCasa() {
        setNumCasas(numCasas + 1);
        return titulo.getPrecioEdificar();
    }

    int edificarHotel() {
        setNumHoteles(numHoteles + 1);
        setNumCasas(numCasas - 4);
        return titulo.getPrecioEdificar();
    }

    boolean estaHipoteca() {
        return titulo.getHipotecada();
    }

    int getCosteHipoteca() {
        return (int) (calcularValorHipoteca() * 1.1);
    }

    int getPrecioEdificar() {
        return titulo.getPrecioEdificar();
    }

    int hipotecar() {
        titulo.setHipotecada(true);
        int cantidadRecibida = this.calcularValorHipoteca();

        return cantidadRecibida;
    }

    boolean sePuedeEdificarCasa(int factorEspeculador) {
        return (numCasas < (4 * factorEspeculador));
    }

    boolean sePuedeEdificarHotel(int factorEspeculador) {
        return (numHoteles < (4 * factorEspeculador) && numCasas == 4);
    }

    boolean tengoPropietario() {
        return titulo.tengoPropietario();
    }

    boolean propietarioEncarcelado() {
        return titulo.propietarioEncarcelado();
    }

    int venderTitulo() {
        this.setNumCasas(0);
        this.setNumHoteles(0);
        titulo.setPropietario(null);
        int precioCompra = super.getCoste() + (numCasas + numHoteles) * titulo.getPrecioEdificar();
        int precioVenta = (int) (precioCompra + precioCompra * titulo.getFactorRevalorizacion());
        return precioVenta;
    }

    private TituloPropiedad asignarTituloPropiedad(Jugador jugador) {
        titulo.setPropietario(jugador);
        return titulo;
    }

    @Override
    boolean soyEdificable() {
        return true;
    }

    @Override
    public String toString() {
        /*String stringCasilla = "Casilla{\nNumero Casilla=" + numeroCasilla + "\nTipo="
                + tipo + "\nCoste=" + coste;*/
        String stringCasilla = super.toString();
        stringCasilla += "\nNumero Casas=" + numCasas
                + "\nNumero Hoteles=" + numHoteles + "\n" + titulo.toString();

        stringCasilla += "\n}";

        return stringCasilla;
    }
}
